File: java/src/main/java/com/leecampbell/cod/domain/contracts/LoanCreatedEvent.java
package com.leecampbell.cod.domain.contracts;

import java.math.BigDecimal;
import java.time.OffsetDateTime;

public final class LoanCreatedEvent extends DomainEvent {
    private final OffsetDateTime createdOn;
    private final Duration term;
    private final PaymentPlan paymentPlan;
    private final BigDecimal amount;

    public LoanCreatedEvent(OffsetDateTime createdOn, Duration term, PaymentPlan paymentPlan, BigDecimal amount) {
        this.createdOn = createdOn;
        this.term = term;
        this.paymentPlan = paymentPlan;
        this.amount = amount;
    }

    public OffsetDateTime createdOn() {
        return createdOn;
    }

    public Duration term() {
        return term;
    }

    public PaymentPlan paymentPlan(){
        return paymentPlan;
    }

    public BigDecimal amount() {
        return amount;
    }
}

File: typescript/packages/domain/src/contracts/loan-created-event.ts
import { Duration } from "./duration.js";
import { Event } from "./event.js";
import { PaymentPlan } from "./payment-plan.js";

export class LoanCreatedEvent extends Event {
  readonly eventType = "LoanCreatedEvent" as const;
  readonly createdOn: Date;
  readonly amount: number;
  readonly term: Duration;
  readonly paymentPlan: PaymentPlan;

  constructor(
    createdOn: Date,
    amount: number,
    term: Duration,
    paymentPlan: PaymentPlan,
  ) {
    super();
    this.createdOn = createdOn;
    this.amount = amount;
    this.term = term;
    this.paymentPlan = paymentPlan;
  }
}
